### PR TITLE
crypto/helper: Add secure wipe function

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

Adds a cryptographically secure wipe function to wipe structs with
sensitive data. Works by first casting the pointer to a `volatile`
pointer to ensure that the compiler doesn't optimize the "memset" away.

### Testing procedure

For now it is mostly if this compiles and if the source makes sense

### Issues/PRs references

None

### CODE REVIEW

**Comment by basilfx** [comment]
What about data remanence attacks? We should either list that this isn't taken care of, or implement a more secure wipe (multiple times, random data, memory type dependent).

I guess the first is the easiest option :-)

**Comment by basilfx** [comment]
What about data remanence attacks? We should either list that this isn't taken care of, or implement a more secure wipe (multiple times, random data, memory type dependent).

I guess the first is the easiest option :-)

**Comment by bergzand** [comment]
Yeah, let's go with the first option. 

Most crypto libraries I know of handle a "secure" wipe this way and only protect against normal memory reads and not against other memory black magic :)

**Comment by bergzand** [comment]
I'm thinking of renaming it to `crypto_secure_wipe` to prevent name collisions with at least Monocypher. Other suggestions are welcome

Marking as WIP to prevent merging before the rename

**Comment by bergzand** [comment]
@basilfx I've added a small note to the function. I've also renamed the function to hopefully avoid naming collision.

File: sys/include/crypto/helper.h [hunk 1]
```diff
@@ -50,12 +50,18 @@ void crypto_block_inc_ctr(uint8_t block[16], int L);
 int crypto_equals(uint8_t *a, uint8_t *b, size_t len);
 
 /**
- * @brief   Secure wipe function
+ * @brief   Secure wipe function.
+ *
+ * This wipe function zeros the supplied buffer in a way that the compiler is
+ * not allowed to optimize. This can be used to erase secrets from memory.
+ *
+ * Note that this function on its owm could be insufficient. It is outside the
```
Comment (basilfx) [inline sys/include/crypto/helper.h]: Typo ;-)

Comment (basilfx) [inline sys/include/crypto/helper.h]: maybe add:

`.. could be insufficient against (data remanence) attacks.`

This clarifies why it is insufficient.

Comment (bergzand) [reply sys/include/crypto/helper.h]: Fixed, thanks for the suggestion

Comment (bergzand) [reply sys/include/crypto/helper.h]: Fixed!

**Comment by basilfx** [comment]
I'm OK with the changes, looks good. Haven't tested it though.

**Comment by bergzand** [comment]
I can add a quick unittest for this, only issue is that verifying the result is a read operation which also  prevents the compiler from optimizing the wipe away. But it would test that the memory region is zero'd out.

**Comment by basilfx** [comment]
Would make sense, since a unit test verifies the code, not the compiler specifics ;-)

**Comment by bergzand** [comment]
Added a simple test to the crypto unittest.

**Comment by bergzand** [comment]
> It is not being optimized away (at least not now).

@jcarrano Thanks for checking. Is this with or without the test assert reading the values after the call?

**Comment by jcarrano** [comment]
Oops! It was with the read. I commented it and dissassembled it again. The code is still there, so it is still secure.

It is declared as volatile, so it *should* get written eventually. The user just needs to be aware of what reorderings are possible (there is a C standard for that).

**Comment by bergzand** [comment]
@basilfx @jcarrano Ready to squash here?

**Comment by basilfx** [comment]
I would say yes. I fully trust @jcarrano checks!

**Comment by bergzand** [comment]
squashed!

**Comment by bergzand** [comment]
@jcarrano Mind pressing the big green button here? :)

**Comment by bergzand** [comment]
@jcarrano Thanks!

**Pull request merged**